pyro/infer/renyi_elbo.py

```diff
@@ -0,0 +1,244 @@
+                        log_prob_sum = log_prob.detach().reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = torch_item(site["log_prob_sum"])
+
+                    elbo_particle = elbo_particle - log_prob_sum
+
+            elbo_trace.append(elbo_particle)
+
+        if self.num_particles == 1:
+            elbo = elbo_trace[0]
+        else:
+            if is_vectorized:
+                elbo_particles = elbo_trace[0]
+            else:
+                elbo_particles = torch.tensor(elbo_trace)  # no need to use .new*() here
+
+            elbo_particles_scaled = (1. - self.alpha) * elbo_particles
+            elbo_scaled = log_sum_exp(elbo_particles_scaled, dim=0) - math.log(self.num_particles)
+            elbo = elbo_scaled.sum().item() / (1. - self.alpha)
+
+        loss = -elbo
+        warn_if_nan(loss, "loss")
+        return loss
+
+    def loss_and_grads(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns an estimate of the ELBO
+        :rtype: float
+
+        Computes the ELBO as well as the surrogate ELBO that is used to form the gradient estimator.
+        Performs backward on the latter. Num_particle many samples are used to form the estimators.
+        """
+        elbo_trace = []
+        surrogate_elbo_trace = []
+        tensor_holder = None
+        is_vectorized = self.vectorize_particles and self.num_particles > 1
+
+        # grab a vectorized trace from the generator
+        for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
+            elbo_particle = 0
+            surrogate_elbo_particle = 0
+
+            # compute elbo and surrogate elbo
+            for name, site in model_trace.nodes.items():
+                if site["type"] == "sample":
+                    if is_vectorized:
+                        log_prob_sum = site["log_prob"].reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+                    elbo_particle = elbo_particle + log_prob_sum.detach()
+                    surrogate_elbo_particle = surrogate_elbo_particle + log_prob_sum
+
+            for name, site in guide_trace.nodes.items():
+                if site["type"] == "sample":
+                    log_prob, score_function_term, entropy_term = site["score_parts"]
+                    if is_vectorized:
+                        log_prob_sum = log_prob.reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+
+                    elbo_particle = elbo_particle - log_prob_sum.detach()
+
+                    if not is_identically_zero(entropy_term):
+                        surrogate_elbo_particle = surrogate_elbo_particle - log_prob_sum
+
+                    if not is_identically_zero(score_function_term):
+                        surrogate_elbo_particle = (surrogate_elbo_particle +
```

Comment:
I'm not familiar with the score parts computation, but it does seem like we are not making use of it in the non reparameterized case. Is this expected?

Reply:
Yes, we don't use it in the case "alpha = 0". The REINFORCE way (for non-reparameterized distribution) will not update the guide when "alpha = 0". I searched for literature and found that [Wake-Sleep](https://arxiv.org/abs/1406.2751) algorithm updates the guide **separately** (by maximizing another likelihood function). Their "wake phase to update guide" is equivalent to add this score_function_term to surrogate_elbo_particle (with some magnitude).

How to solve this problem and how to reduce variance for non-reparameterized distributions are good topics for a research paper in my opinion (because I couldn't find any article discussing them).

Reply:
Thanks for explaining, @fehiepsi. Makes sense!